src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1404,7 +1421,20 @@ internal bool HasTopLevelNullabilityImplicitConversion(TypeSymbolWithAnnotations
             {
                 return true;
             }
-            return source.IsNullable != true || destination.IsNullable != false;
+
+            if (source.NullableAnnotation == NullableAnnotation.Unknown ||
+                destination.NullableAnnotation == NullableAnnotation.Unknown ||
+                destination.NullableAnnotation.IsAnyNullable())
+            {
+                return true;
+            }
+
+            if (source.IsPossiblyNullableReferenceTypeTypeParameter() && !destination.IsPossiblyNullableReferenceTypeTypeParameter())
+            {
+                return false;
+            }
+
+            return !source.NullableAnnotation.IsAnyNullable();
```

Comment:
A similar exercise would also be useful for the `HasTopLevelNullabilityImplicitConversion` method.
Two ways I could think to explain it in an xml doc:
1. returns true if you can assign source to destination (in terms of top-level nullability)
2. returns true except if the source is value-nullable but the destination is value-not-nullable

Again, coming to a good conceptual understanding of the method will help with reviewing and possibly also with simplifying the code.

Reply:
> A similar exercise would also be useful for the HasTopLevelNullabilityImplicitConversion method.

Adding a doc-comment in the next commit

---
In reply to: [231458546](https://github.com/dotnet/roslyn/pull/30913#discussion_r231458546) [](ancestors = 231458546)